casr/src/bin/casr-ubsan.rs

```diff
@@ -0,0 +1,456 @@
+                        return Err(err);
+                    }
+                    Ok(i_dir.to_path_buf())
+                })
+        )
+        .arg(
+            Arg::new("output")
+                .short('o')
+                .long("output")
+                .action(ArgAction::Set)
+                .required(true)
+                .value_name("OUTPUT_DIR")
+                .value_parser(clap::value_parser!(PathBuf))
+                .help("Output directory with triaged reports")
+        )
+        .arg(
+            Arg::new("ARGS")
+                .action(ArgAction::Set)
+                .required(false)
+                .num_args(1..)
+                .last(true)
+                .help("Add \"-- <path> <arguments>\" to run"),
+        )
+        .get_matches();
+
+    // Init log.
+    util::initialize_logging(&matches);
+
+    debug!("Get args");
+    // Get input dir list
+    let input_dirs: Vec<_> = matches.get_many::<PathBuf>("input").unwrap().collect();
+    // Get output dir
+    let output_dir = matches.get_one::<PathBuf>("output").unwrap();
+    if !output_dir.exists() {
+        fs::create_dir_all(output_dir).with_context(|| {
+            format!("Couldn't create output directory {}", output_dir.display())
+        })?;
+    } else if !output_dir.is_dir() {
+        bail!("Output directory must be a directory");
+    } else if output_dir.read_dir()?.next().is_some() {
+        bail!("Output directory is not empty.");
+    }
+    // Get program args.
+    let argv: Vec<&str> = if let Some(argvs) = matches.get_many::<String>("ARGS") {
+        argvs.map(|s| s.as_str()).collect()
+    } else {
+        bail!("Wrong arguments for starting program");
+    };
+
+    // Get timeout
+    let timeout = if let Some(timeout) = matches.get_one::<u64>("timeout") {
+        *timeout
+    } else {
+        0
+    };
+
+    // Get input path list
+    debug!("Get input path list");
+    let mut inputs: Vec<PathBuf> = vec![];
+    // Do without paralleling to preserve the specified order
+    for input_dir in input_dirs {
+        for path in WalkDir::new(input_dir)
+            .sort_by_file_name()
+            .into_iter()
+            .filter_map(|file| file.ok())
+            .filter(|file| file.metadata().unwrap().is_file())
+            .map(|file| file.path().to_path_buf())
+        {
+            inputs.push(path);
+        }
+    }
+
+    // Get number of threads
+    debug!("Get number of threads");
+    let jobs = if let Some(jobs) = matches.get_one::<u32>("jobs") {
+        *jobs as usize
+    } else {
+        std::cmp::max(1, num_cpus::get() / 2)
+    };
+    let num_of_threads = jobs.min(inputs.len()).max(1);
+    let custom_pool = rayon::ThreadPoolBuilder::new()
+        .num_threads(num_of_threads)
+        .build()
+        .unwrap();
+
+    // Set ubsan env options
+    debug!("Set environment");
```

Comment:
Remove. Does not provide any useful information.

Reply:
Done